From 0875372a9c51d429b038fbdb47221e7d1835066c Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 13:13:05 -0400
Subject: [ethernet] i40evf: don't leak queue vectors

Message-id: <1409663591-32531-143-git-send-email-sassmann@redhat.com>
Patchwork-id: 90620
O-Subject: [RHEL7.1 PATCH 142/148] i40evf: don't leak queue vectors
Bugzilla: 1091127
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitch Williams <mitch.a.williams@intel.com>

Fix a memory leak. Driver was allocating memory for queue vectors on
init but not freeing them on shutdown. These need to be freed at two
different times: during module unload, and during reset recovery when
the driver cannot contact the PF driver and needs to give up.

Change-ID: I7c1d0157a776e960d4da432dfe309035aad7c670
Signed-off-by: Mitch Williams <mitch.a.williams@intel.com>
Tested-by: Sibai Li <sibai.li@intel.com>
Signed-off-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit d31944d6f0872ba9536501c4850dfb20dea5bf64)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40evf/i40evf_main.c b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
index 2aca9cf..0c4f7ba 100644
--- a/drivers/net/ethernet/intel/i40evf/i40evf_main.c
+++ b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
@@ -1542,6 +1542,7 @@ static void i40evf_reset_task(struct work_struct *work)
   i40evf_free_misc_irq(adapter);
   i40evf_reset_interrupt_capability(adapter);
   i40evf_free_queues(adapter);
+  i40evf_free_q_vectors(adapter);
   kfree(adapter->vf_res);
   i40evf_shutdown_adminq(hw);
   adapter->netdev->flags &= ~IFF_UP;
@@ -2429,6 +2430,7 @@ static void i40evf_remove(struct pci_dev *pdev)
   i40evf_misc_irq_disable(adapter);
   i40evf_free_misc_irq(adapter);
   i40evf_reset_interrupt_capability(adapter);
+  i40evf_free_q_vectors(adapter);
  }
 
  if (adapter->watchdog_timer.function)
-- 
1.7.1